src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,321 @@
+                                                                                                                    s_localizableDescription,
+                                                                                                                    isPortedFxCopRule: false,
+                                                                                                                    isDataflowRule: false);
+        internal static DiagnosticDescriptor ImplementsEmptyPreviewInterfaceRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_implementsEmptyPreviewInterface,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildError,
+                                                                                                                    s_localizableDescription,
+                                                                                                                    isPortedFxCopRule: false,
+                                                                                                                    isDataflowRule: false);
+
+        private const string RequiresPreviewFeaturesAttribute = nameof(RequiresPreviewFeaturesAttribute);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralPreviewFeatureAttributeRule, StaticAbstractIsPreviewFeatureRule, ImplementsEmptyPreviewInterfaceRule);
+
+        private enum PreviewFeatureUsageType
+        {
+            General = 0,
+            StaticAbstract,
+            EmptyInterface
+        }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeVersioningRequiresPreviewFeaturesAttribute, out var previewFeaturesAttribute))
+                {
+                    return;
+                }
+
+                Compilation compilation = context.Compilation;
+                IAssemblySymbol? thisAssembly = compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
```

Comment:
NIT: IMO these variables are redundant as not used anywhere else, but if you prefer this way it's OK, up to you
```suggestion
                AttributeData? attribute = context.Compilation.Assembly.GetAttributes().FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
```